### PR TITLE
Added config support to the mod.

### PR DESCRIPTION
**Pull request by 3therios**
Added config functionality for all the configurable properties of a Logistic stations.
Added some utility methods for managing all the attributes.
Also address issue #1 in the second commit.

TODO: Determine how to set the trip range for drones to anything other than 180°.
Right now it's hardcoded to -1. Any help on this, if you understand how it works would be appreciated @Pasukaru.

I can be reached on Discord (3therios#0794)

### CODE REVIEW

**Comment by alpe12** [comment]
Awesome! Thanks.
Can you add a config for the amount of drones/vessels that gets added from inventory? Or a simple enable/disable would be fine.

**Comment by 3therios** [comment]
> Awesome! Thanks.
> Can you add a config for the amount of drones/vessels that gets added from inventory? Or a simple enable/disable would be fine.

Have added that.

**Comment by 3therios** [comment]
> Awesome! Thanks.
> Can you add a config for the amount of drones/vessels that gets added from inventory? Or a simple enable/disable would be fine.

Donezo

**Comment by Pasukaru** [comment]
Cool, thanks for the PR! 
Looks good on first glance.
Will look into it later to merge and publish a release!

**Comment by Pasukaru** [comment]
A few things I noticed: Amount of drones, amount of power config limits should not be absolute, rather percentages of the station.

This is to have support for modded stations, where the max power output or max amount drones/vessels is different from the vanilla ones.

**Comment by Pasukaru** [comment]
> TODO: Determine how to set the trip range for drones to anything other than 180°.
> Right now it's hardcoded to -1. Any help on this, if you understand how it works would be appreciated @Pasukaru.

Didn't take a deeper look into it. The only thing that I found out while playing around with it was that the game uses -1 as infinite in this game. Will try to find out the formula in the evening to add support for this config variable.

**Comment by 3therios** [comment]
> This is to have support for modded stations, where the max power input or max amount drones/vessels is different from the vanilla ones (it may be higher, or even lower).
> 
> What do you think?

Sounds like a good idea.

> A few things I noticed: Amount of drones, amount of power config values and limits should not be absolute, rather percentages of the station.

The auto load of vessels and drones should already support this in part since it has a max load property.
So maybe not percentage but still be configurable to be > 50 or 10 like it is in vanilla. 

With the power side of things, I'm not 100% sure how it works so might need some guidance on that.

Contextually, I'm quite newbie to the mod development / unity and just saw an opportunity for an easy trial attempt.
I just saw that you were multiplying it by 5 to get the 300 MW so not sure if there's a way to determine the max load available from a power station.

**Comment by Pasukaru** [comment]
>The auto load of vessels and drones should already support this in part since it has a max load property.
So maybe not percentage but still be configurable to be > 50 or 10 like it is in vanilla.

I still prefer a percentage based value here. The player can configure 0 if they want to disable the auto-insert, or 100 to insert however many the station accepts. Or anything in between.
Not sure if there is an easy way to distinguish modded stations from vanilla ones. Then we could add overrides to the config. 

Something alone those lines:
```
General.PLS.droneInsertPercent: 100
General.PLS.<modded identifier>.droneInsertAbsolute: 150
```
If there is no override, we fall back to the vanilla percentages.

As a first implementation, I'd just go with the percentages, and move the modded station detection to later.

> I just saw that you were multiplying it by 5 to get the 300 MW so not sure if there's a way to determine the max load available from a power station.

I copied the logic from what the game does. It internally stores the the value X, and to display it as MW in the ui on the sliders it multiplies it by 5. 
It does similar things to other values, so it always involves a little bit of digging through the UI code to figure which conversion happens where. Unfortunately, they don't have it in separate utility classes that we could reuse.

For consistency, I'd display the values in the config like the game does in the UI, since the player has no idea about the underlying data structure.

**Comment by 3therios** [comment]
> > The auto load of vessels and drones should already support this in part since it has a max load property.
> > So maybe not percentage but still be configurable to be > 50 or 10 like it is in vanilla.
>
> I still prefer a percentage based value here. The player can configure 0 if they want to disable the auto-insert, or 100 to insert however many the station accepts. Or anything in between.
> Not sure if there is an easy way to distinguish modded stations from vanilla ones. Then we could add overrides to the config.
> 
> Something alone those lines:
> 
> ```
> General.PLS.droneInsertPercent: 100
> General.PLS.<modded identifier>.droneInsertAbsolute: 150
> ```
> 
> If there is no override, we fall back to the vanilla percentages.
> 
> As a first implementation, I'd just go with the percentages, and move the modded station detection to later.
>

That seems do-able. Given we have a maximum storage value we can easily determine the equivalent load based on percentages. I'll update that now for it to support 0 - 100 and calculate the percentage equivalent to load in (closest int)
 
> > I just saw that you were multiplying it by 5 to get the 300 MW so not sure if there's a way to determine the max load available from a power station.
> 
> I copied the logic from what the game does. It internally stores the the value X, and to display it as MW in the ui on the sliders it multiplies it by 5.
> It does similar things to other values, so it always involves a little bit of digging through the UI code to figure which conversion happens where. Unfortunately, they don't have it in separate utility classes that we could reuse.
> 
> For consistency, I'd display the values in the config like the game does in the UI, since the player has no idea about the underlying data structure.

So you thinking the actual value 30 to 300?

**Comment by Pasukaru** [comment]
> So you thinking the actual value 30 to 300?

No, it is a higher value. Don't know the exact one off the top of my head. Maybe the game stores it in W or even something else.

For AU & LY, I already did find it out, you can find it in the Util.cs: https://github.com/Pasukaru/DSP-Mods/blob/main/AutoStationConfig/Util.cs

>  I copied the logic from what the game does. It internally stores the the value X, and to display it as MW in the ui on the sliders it multiplies it by 5.
    It does similar things to other values, so it always involves a little bit of digging through the UI code to figure which conversion happens where. Unfortunately, they don't have it in separate utility classes that we could reuse.

Correction: It seems like there is no maximum value per-se. It's just that the game's UI code computes the min value for the power slider as X/2, and the maximum value as X*5. This is why I copied the * 5 in `prefabDesc.workEnergyPerTick * 5L`, into my implementation to set it to this artificial max. Via code, you could set it even higher than that, also works with drones - you can add more than the max would allow. Probably not a good idea to do that though :).

To sum up, we would need a similar conversion method for absolute MW values. 
However, I'd still like to go with percentages here as well. So we won't even need to find the conversion for power units, since the user would insert 10% in the config, and we calculate that from the artificial max (`prefabDesc.workEnergyPerTick * 5L * configPercent`). 

So, the game UI would do the rest. Input validation should only accept those percentages that the UI slider allows, not cause any inconsistencies. Which I believe is just starting and 10%, with increments of 10% up to 100%. 
Would also set our config defaults to 100 here.

**Comment by 3therios** [comment]
> > So you thinking the actual value 30 to 300?
> 
> No, it is a higher value. Don't know the exact one off the top of my head. Maybe the game stores it in W or even something else.

Just did a bit of testing. The game stores the charge rate as joules, i.e. joules per tick. The default is 1,000,000 which apparently translates to 60 MW.

> 
> For AU & LY, I already did find it out, you can find it in the Util.cs: https://github.com/Pasukaru/DSP-Mods/blob/main/AutoStationConfig/Util.cs

Have noticed and used this. Didn't actually realise it was something you defined till just then haha.

> > I copied the logic from what the game does. It internally stores the the value X, and to display it as MW in the ui on the sliders it multiplies it by 5.
> > It does similar things to other values, so it always involves a little bit of digging through the UI code to figure which conversion happens where. Unfortunately, they don't have it in separate utility classes that we could reuse.

Not a 100% how to dig through the UI code TBH.

> Correction: It seems like there is no maximum value per-se. It's just that the game's UI code computes the min value for the power slider as X/2, and the maximum value as X*5. This is why I copied the * 5 in `prefabDesc.workEnergyPerTick * 5L`, into my implementation to set it to this artificial max. Via code, you could set it even higher than that, also works with drones - you can add more than the max would allow. Probably not a good idea to do that though :).
> 
> To sum up, we would need a similar conversion method for absolute MW values.
> However, I'd still like to go with percentages here as well. So we won't even need to find the conversion for power units, since the user would insert 10% in the config, and we calculate that from the artificial max (`prefabDesc.workEnergyPerTick * 5L * configPercent`).
> 
> So, the game UI would do the rest. Input validation of the config file should only accept those percentages that the UI slider allows, to not cause any inconsistencies. Which I believe is just starting at 10%, with increments of 10% up to 100%.
> Would also set our config defaults to 100 here.

The slider increment appears to be based on 1% increments of the 300 max, with a minimum of 10%. So we can enforce 10% - 100% config support? The expression you've provided above should work fine to calculate it. I've just been working on it, so I'll push that logic up in a tick.

**Comment by 3therios** [comment]
@Pasukaru - Just added the percent based power and loading of drones / vessels.
LMK if there's anything else.
HMU on Discord if you'd like to chat more. Would love to get an understanding on how you're looking at the UI code and determining what happens and also how you debug the mod / code during development.
So far using the Logger to file has been less than ideal and inefficient, but does work to some extent, but I have no idea how you find the appropriate classes and things. New to C# and Unity all around.

**Comment by Pasukaru** [comment]
@3therios 
Didn't have permisssions to push to your branch, so i continued here: #3.

Anything you'd like to add?
